patches/react-native-gradle-plugin+0.0.7.patch

```diff
@@ -0,0 +1,38 @@
+diff --git a/node_modules/react-native-gradle-plugin/build.gradle.kts b/node_modules/react-native-gradle-plugin/build.gradle.kts
+index c457a57..e1d46a8 100644
+--- a/node_modules/react-native-gradle-plugin/build.gradle.kts
++++ b/node_modules/react-native-gradle-plugin/build.gradle.kts
+@@ -9,7 +9,7 @@ import org.gradle.api.internal.classpath.ModuleRegistry
+ import org.gradle.configurationcache.extensions.serviceOf
+ 
+ plugins {
+-  kotlin("jvm") version "1.6.10"
++  kotlin("jvm") version "1.8.0"
+   id("java-gradle-plugin")
+ }
+ 
+@@ -31,7 +31,7 @@ group = "com.facebook.react"
+ 
+ dependencies {
+   implementation(gradleApi())
+-  implementation("com.android.tools.build:gradle:7.0.4")
++  implementation("com.android.tools.build:gradle:7.2.1")
+   implementation("com.google.code.gson:gson:2.8.9")
+   implementation("com.google.guava:guava:31.0.1-jre")
+   implementation("com.squareup:javapoet:1.13.0")
+@@ -46,12 +46,12 @@ dependencies {
+ }
+ 
+ java {
+-    sourceCompatibility = JavaVersion.VERSION_1_8
+-    targetCompatibility = JavaVersion.VERSION_1_8
++    sourceCompatibility = JavaVersion.VERSION_17
++    targetCompatibility = JavaVersion.VERSION_17
+ }
+ 
+ tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
+     kotlinOptions {
+-        jvmTarget = JavaVersion.VERSION_11.majorVersion
++        jvmTarget = JavaVersion.VERSION_17.majorVersion
```

Comment:
1. Question (❓): I am not sure why and how this patch got generated, where did you get this newer version of this file? 🤔 For example, I am seeing that the newest version of such [react-native-gradle-plugin](https://github.com/invertase/react-native-gradle-plugin/) is [this](https://github.com/invertase/react-native-gradle-plugin/blob/master/build.gradle).
2. If that was an arbitrary update, that is, just to get this working:
    1. Question (❓): Was updating `Kotlin` necessary, and if so, why `1.8.10`?
    2. Question (❓): Was updating `AGP` necessary, and if so, why `7.2.1`?
    3. Warning (⚠️): Why Java `sourceCompatibility`, `targetCompatibility` and Kotlin `jvmTarget` are all now pointing to Java 17 (aka `JavaVersion.VERSION_17`)? I am asking that because I do believe this will cause trouble on older versions of Android. Instead, and if at all possible, I recommend all of them to be targeting Java 8 (aka `JavaVersion.VERSION_1_8`).

Reply:
> 1. Question (❓): I am not sure why and how this patch got generated, where did you get this newer version of this file? 🤔

This patch was generated using the [`patch-package` NPM command](https://www.npmjs.com/package/patch-package). It basically generates a patch and applies it every time you install the dependencies. It's a good alternative when you need to tweak something from a package but you don't want to fork it.

> For example, I am seeing that the newest version of such [react-native-gradle-plugin](https://github.com/invertase/react-native-gradle-plugin/) is [this](https://github.com/invertase/react-native-gradle-plugin/blob/master/build.gradle).

I think that's a different plugin, the one we are using comes from the React Native repository:
https://github.com/facebook/react-native/tree/HEAD/packages/react-native-gradle-plugin

In fact, I'm thinking that we might reference a newer version of this package to solve the issue 🤔.

> * Question (❓): Was updating `Kotlin` necessary, and if so, why `1.8.10`?

I updated it to solve the following error:
```
e: Incompatible classes were found in dependencies. Remove them from the classpath or use '-Xskip-metadata-version-check' to suppress errors
e: .gradle/caches/8.2.1/generated-gradle-jars/gradle-api-8.2.1.jar!/META-INF/configuration-cache.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/caches/8.2.1/generated-gradle-jars/gradle-api-8.2.1.jar!/META-INF/platform-jvm.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/META-INF/kotlin-stdlib-jdk7.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/META-INF/kotlin-stdlib-jdk8.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/META-INF/kotlin-stdlib.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-common-1.8.20.jar!/META-INF/kotlin-stdlib-common.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: workspace/gutenberg-mobile/gutenberg/node_modules/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/AndroidConfiguration.kt: (19, 25): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
The class is loaded from .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/kotlin/Unit.class
```

> * Question (❓): Was updating `AGP` necessary, and if so, why `7.2.1`?

I've tested again reverting this change and the build was successful. Probably I update it while testing and forgot to change it back, I'll revert it.

> * Warning (⚠️): Why Java `sourceCompatibility`, `targetCompatibility` and Kotlin `jvmTarget` are all now pointing to Java 17 (aka `JavaVersion.VERSION_17`)? I am asking that because I do believe this will cause trouble on older versions of Android. Instead, and if at all possible, I recommend all of them to be targeting Java 8 (aka `JavaVersion.VERSION_1_8`).

Ah, interesting. I changed it to solve the following error:
```
java.lang.IllegalArgumentException: Unknown Kotlin JVM target: 8
```

I noticed in the React Native repository that they changed this value to `VERSION_11` ([reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L53-L54)). Looks like the main reason was upgrading Kotlin to `1.8.0` and the fact that Kotlin changed the default version of the stdlib that ships with by default.

<hr>

I'm going to try to use a newer version of the `react-native-gradle` package, as seems the optimal approach.

Reply:
Regarding issues with `react-native-gradle-plugin`, my understanding is that using Gradle 8 requires also bump Kotlin to 1.8. Note that the version of the Gradle plugin that RN 0.69.4 uses points to an older version 1.6 ([reference](https://github.com/facebook/react-native/blob/4bdec975c95e8c46c804a8a00c73ebe696a6982b/packages/react-native-gradle-plugin/build.gradle.kts#L12)), this is likely producing the issue I shared on the first question.

What it's not totally clear to me is why we need to use Java 17, seems related to the fact that we use AGP 8 but I'm not totally sure 🤔

Reply:
I tested using the latest version of the React Native Gradle plugin (package `@react-native/gradle-plugin`) and the Gutenberg demo app is building successfully. I think for now we can take this approach (https://github.com/WordPress/gutenberg/pull/52872/commits/693c0fe639dc8bf116ab4398782c74f2bf29766e).

Reply:
👋 @fluiddot !

> This patch was generated using the [patch-package NPM command](https://www.npmjs.com/package/patch-package). It basically generates a patch and applies it every time you install the dependencies. It's a good alternative when you need to tweak something from a package but you don't want to fork it.

TIL about [patch-package NPM command](https://www.npmjs.com/package/patch-package) and thanks for the clarification on that! 🙏 

> I think that's a different plugin, the one we are using comes from the React Native repository:
https://github.com/facebook/react-native/tree/main/packages/react-native-gradle-plugin

Oh, another TIL about [facebook -> react-native-gradle-plugin](https://github.com/facebook/react-native/tree/main/packages/react-native-gradle-plugin) and many thanks for helping me see the RN light! 🙏 

> In fact, I'm thinking that we might reference a newer version of this package to solve the issue 🤔.

🤔 

> I updated it to solve the following error:

Thanks for the clarification on that! 🙏 

> I've tested again reverting this change and the build was successful. Probably I update it while testing and forgot to change it back, I'll revert it.

👍 but also, I was actually thinking if we could use `8.0.2`, probably not, right? 🤔 

> Ah, interesting. I changed it to solve the following error:

👍 

> I noticed in the React Native repository that they changed this value to VERSION_11 ([reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L53-L54)). Looks like the main reason was upgrading Kotlin to 1.8.0 and the fact that Kotlin changed the default version of the stdlib that ships with by default.

🤔 Yea, this [Java reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L53-L54) and this [Kotlin reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L59) are both pointing to `VERSION_11`, but still, I am not sure what that means for us. One the one hand we won't be using any `VERSION_11`, it seems that it is just that the plugin is using that, but on the other hand, everywhere else we are still using `VERSION_1_8`, thus, I bit confused on what's the correct action here, it seems using `VERSION_11` like that is okay, but I can't be sure, needs testing... 🤷 

> I'm going to try to use a newer version of the react-native-gradle package, as seems the optimal approach.

👍 

> Regarding issues with react-native-gradle-plugin, my understanding is that using Gradle 8 requires also bump Kotlin to 1.8.

As far as I am aware, Gradle 8 doesn't have such a requirement. For reference, check Aztec, which is on [Kotlin 1.6.10](https://github.com/wordpress-mobile/AztecEditor-Android/blob/trunk/settings.gradle#L2), and it can still be upgraded to Gradle 8 (see [relevant PR](https://github.com/wordpress-mobile/AztecEditor-Android/pull/1051)).

> Note that the version of the Gradle plugin that RN 0.69.4 uses points to an older version 1.6 ([reference](https://github.com/facebook/react-native/blob/4bdec975c95e8c46c804a8a00c73ebe696a6982b/packages/react-native-gradle-plugin/build.gradle.kts#L12)), this is likely producing the issue I shared on the first question.

👍 

> What it's not totally clear to me is why we need to use Java 17, seems related to the fact that we use AGP 8 but I'm not totally sure 🤔

Yeap, not sure as well, but also, I don't think this is related to AGP 8 (see `Aztec` example above).

> I tested using the latest version of the React Native Gradle plugin (package @react-native/gradle-plugin) and the Gutenberg demo app is building successfully. I think for now we can take this approach (https://github.com/WordPress/gutenberg/commit/693c0fe639dc8bf116ab4398782c74f2bf29766e).

Awesome and yes, this is indeed seems the safer choice here, no need for us to decide on Java, Kotlin and AGP version and compatibility explicitly, thus potentially making a mess out of it! 🎉 

-----

🙇 ❤️ 🚀 THANK YOU 🙇 ❤️ 🚀